runelite-client/src/main/java/net/runelite/client/plugins/timetracking/farming/FarmingTracker.java

```diff
@@ -457,4 +474,112 @@ else if (maxCompletionTime <= Instant.now().getEpochSecond())
 			completionTimes.put(tab.getKey(), completionTime);
 		}
 	}
+
+	public void checkCompletion()
+	{
+		List<RuneScapeProfile> rsProfiles = configManager.getRSProfiles();
+		long unixNow = Instant.now().getEpochSecond();
+
+		for (RuneScapeProfile profile : rsProfiles)
+		{
+			Integer offsetPrecisionMins = configManager.getConfiguration(TimeTrackingConfig.CONFIG_GROUP, profile.getKey(), TimeTrackingConfig.FARM_TICK_OFFSET_PRECISION, int.class);
+			Integer offsetTimeMins = configManager.getConfiguration(TimeTrackingConfig.CONFIG_GROUP, profile.getKey(), TimeTrackingConfig.FARM_TICK_OFFSET, int.class);
+
+			RuneScapeProfileType profileType = profile.getType();
+
+			for (Map.Entry<Tab, Set<FarmingPatch>> tab : farmingWorld.getTabs().entrySet())
+			{
+				for (FarmingPatch patch : tab.getValue())
+				{
+					String configKey = NOTIFY_CONFIG_KEY +
+						"." + patch.getRegion().getRegionID() +
+						"." + patch.getVarbit().getId();
+					boolean shouldNotify = Boolean.TRUE
+						.equals(configManager.getConfiguration(TimeTrackingConfig.CONFIG_GROUP, profile.getKey(), configKey, Boolean.class));
+
+					if (shouldNotify)
+					{
+						PatchPrediction prediction = predictPatch(patch, profile.getKey());
+						if (prediction != null)
+						{
+							int tickRate = prediction.getProduce().getTickrate();
+
+							//Only notify if recorded offset is large enough for the current patch
+							if (offsetPrecisionMins != null && offsetTimeMins != null && (offsetPrecisionMins >= tickRate || offsetPrecisionMins >= 40))
+							{
+								if ((unixNow - 10 <= prediction.getDoneEstimate() && prediction.getDoneEstimate() <= unixNow) && prediction.getProduce() != Produce.WEEDS)
```

Comment:
My only other thought was for each time the method is fired record the Unix time and compare to that instead of `unixNow-10`.

Reply:
what you need to avoid is:
1) the scheduler being blocked for a long period of time, and thus not calling this exactly on the right time, should not drop notifications
2) Instant.now() is not monotonic, so time moving backwards (or forwards) should not drop/duplicate notifications

Probably the most reliable thing would be to record the state of the crops as of the previous check and only fire notifications if the state has changed

Reply:
I'm not sure how to do this as while predictPatch returns a CropState, it's just the initial CropState of the varbit it reads from the config (ie the last observed state), so we'd still need some method at some point to compare predictPatch's `doneEstimate` with the current time to update the CropState to `HARVESTABLE`